File: frontend/src/lib/rawMaterials.ts
// Raw-material cost-trend mapping for the company page.
//
// The backend bakes raw_materials.json: per-commodity trend (from real futures
// moving averages) + a sector→commodity map + an aggregate outlook. Here we attach
// the right commodities to a company by its sector/industry and translate each
// input's trend into a margin impact — a company that *consumes* an input benefits
// when that input's price falls. Deterministic, not a forecast.

import type { CommoditySummary } from "@/lib/research";

export type Trend = "increasing" | "decreasing" | "sideways";

export interface Commodity {
  symbol: string;
  name: string;
  trend: Trend;
  change_pct: number | null;
  close: number | null;
}

export interface RawMaterialsData {
  commodities: Record<string, Commodity>;
  sector_map: { keywords: string[]; materials: string[] }[];
  outlook: string;
  counts: { decreasing: number; increasing: number; sideways: number };
}

export interface MaterialImpact {
  name: string;
  symbol: string;
  trend: Trend;
  changePct: number | null;
  impact: "Positive" | "Neutral" | "Negative";
  effect: string;
}

/** Consumer-side rule: falling input price → positive margin impact. */
function impactFor(trend: Trend): { impact: MaterialImpact["impact"]; effect: string } {
  if (trend === "decreasing")
    return { impact: "Positive", effect: "Lower input cost — supports gross-margin expansion" };
  if (trend === "increasing")
    return { impact: "Negative", effect: "Rising input cost — pressures gross margins" };
  return { impact: "Neutral", effect: "Stable input cost — margins broadly unchanged" };
}

/** Commodities relevant to a company, resolved by sector/industry keywords. */
export function companyMaterials(
  sector: string | null,
  industry: string | null,
  data: RawMaterialsData | null,
): MaterialImpact[] {
  if (!data) return [];
  const hay = `${sector ?? ""} ${industry ?? ""}`.toLowerCase();
  const symbols = new Set<string>();
  for (const entry of data.sector_map) {
    if (entry.keywords.some((k) => hay.includes(k))) {
      for (const m of entry.materials) symbols.add(m);
    }
  }
  const out: MaterialImpact[] = [];
  for (const sym of symbols) {
    const c = data.commodities[sym];
    if (!c) continue;
    const { impact, effect } = impactFor(c.trend);
    out.push({ name: c.name, symbol: sym, trend: c.trend, changePct: c.change_pct, impact, effect });
  }
  return out;
}

export function commoditySummary(materials: MaterialImpact[]): CommoditySummary {
  const hasInputs = materials.length > 0;
  const anyIncreasing = materials.some((m) => m.trend === "increasing");
  const decreasing = materials.filter((m) => m.trend === "decreasing").length;
  return { hasInputs, anyIncreasing, favorable: hasInputs && decreasing > materials.length / 2 };
}

File: backend/app/services/raw_materials.py
"""Raw-material cost-trend intelligence for the company research page.

Professional investors watch input costs *before* results land. We already ingest
liquid commodity futures (Yahoo ``=F``); this module classifies each commodity's
current trend from its own moving averages (no prediction, just direction of the
real series) and maps company sectors to the commodities that are genuine cost
drivers. The per-company margin impact is a deterministic rule — a company that
*consumes* an input benefits when that input's price falls — not a forecast.

Output is baked into ``raw_materials.json`` (global, sector-keyed) so the frontend
can attach the right materials to any company by its sector without a per-company
recompute.
"""

from __future__ import annotations

import json
from pathlib import Path

# Tracked commodity symbol → display name. Only liquid futures we actually ingest.
COMMODITY_NAMES: dict[str, str] = {
    "GC": "Gold", "SI": "Silver", "PL": "Platinum", "PA": "Palladium",
    "HG": "Copper", "CL": "Crude Oil (WTI)", "BZ": "Brent Crude",
    "NG": "Natural Gas", "RB": "Gasoline", "HO": "Heating Oil",
    "ZC": "Corn", "ZW": "Wheat", "ZS": "Soybeans", "KC": "Coffee",
    "SB": "Sugar", "CC": "Cocoa", "CT": "Cotton", "LE": "Live Cattle",
}

# Sector/industry keyword (lower-case, substring match) → cost-driver commodities,
# most-relevant first. Keys cover GICS (US/curated) and PSX portal sector names.
SECTOR_INPUTS: list[tuple[tuple[str, ...], list[str]]] = [
    (("cement",), ["NG", "CL"]),
    (("fertil",), ["NG"]),
    (("chemical", "petrochemical"), ["CL", "NG"]),
    (("textile", "spinning", "synthetic", "rayon", "apparel"), ["CT", "CL"]),
    (("sugar",), ["SB"]),
    (("food", "personal care", "consumer staples", "consumer defensive",
      "beverage", "tobacco"), ["ZW", "ZC", "SB", "ZS"]),
    (("agri", "agriculture"), ["ZC", "ZW", "ZS"]),
    (("oil & gas", "oil and gas", "petroleum", "refinery", "energy"), ["CL", "NG"]),
    (("power", "utilit"), ["NG", "CL"]),
    (("glass", "ceramic"), ["NG"]),
    (("automobile", "auto ", "auto,", "autos", "vehicle"), ["HG", "CL"]),
    (("engineering", "steel", "iron", "metal"), ["HG"]),
    (("airline", "transport", "logistics", "shipping", "aviation"), ["CL"]),
    (("paper", "board", "packaging"), ["NG"]),
    (("materials",), ["HG", "CL"]),
]


def classify_trend(close: float | None, sma50: float | None, sma200: float | None) -> str:
    """Direction of the real price series from its moving averages."""
    if close and sma50 and sma200:
        if close < sma50 < sma200:
            return "decreasing"
        if close > sma50 > sma200:
            return "increasing"
    if close and sma50:
        if close < sma50 * 0.98:
            return "decreasing"
        if close > sma50 * 1.02:
            return "increasing"
    return "sideways"


def _read_commodity(company_dir: Path, symbol: str) -> dict | None:
    """Read a commodity company JSON (provider_symbol is ``SYMBOL=F``)."""
    path = company_dir / f"{symbol}=F.json"
    if not path.exists():
        return None
    try:
        d = json.loads(path.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError):
        return None
    q = d.get("quote") or {}
    t = d.get("technical") or {}

    def num(v):
        try:
            return float(v)
        except (TypeError, ValueError):
            return None

    close = num(q.get("price"))
    trend = classify_trend(close, num(t.get("sma_50")), num(t.get("sma_200")))
    return {
        "symbol": symbol,
        "name": COMMODITY_NAMES.get(symbol, symbol),
        "trend": trend,
        "change_pct": num(q.get("change_pct")),
        "close": close,
    }


def build_raw_materials(company_dir: Path) -> dict:
    """Global raw-material trend map, built from committed commodity JSONs."""
    commodities: dict[str, dict] = {}
    for sym in COMMODITY_NAMES:
        row = _read_commodity(company_dir, sym)
        if row is not None:
            commodities[sym] = row

    sector_map = [
        {"keywords": list(keys), "materials": [m for m in mats if m in commodities]}
        for keys, mats in SECTOR_INPUTS
    ]
    sector_map = [s for s in sector_map if s["materials"]]

    # Aggregate outlook from how many tracked inputs are falling vs rising.
    trends = [c["trend"] for c in commodities.values()]
    down = trends.count("decreasing")
    up = trends.count("increasing")
    if down > up + 2:
        outlook = ("Input-cost environment is broadly improving — a majority of "
                   "tracked commodity inputs are declining, which is supportive of "
                   "gross margins across cost-sensitive sectors.")
    elif up > down + 2:
        outlook = ("Input-cost environment is worsening — a majority of tracked "
                   "commodity inputs are rising, which pressures gross margins in "
                   "cost-sensitive sectors.")
    else:
        outlook = ("Input-cost environment is mixed — commodity input trends are "
                   "divergent, so margin impact is company- and sector-specific.")

    return {
        "commodities": commodities,
        "sector_map": sector_map,
        "outlook": outlook,
        "counts": {"decreasing": down, "increasing": up, "sideways": trends.count("sideways")},
    }
